File: 프로그래머스/2/12909. 올바른 괄호/올바른 괄호.js
function solution(s){
    const stack = [];
    
    for (char of s.split("")) {
        if (stack.length === 0 && char === ')') {
            return false;
        }
        else if (char === '(') stack.push(char)
        else stack.pop();
    }
    
    if (stack.length > 0) return false;
    
    return true;
}


// stack에 '('일 때는 요소를 삽입 / ')'일 때는 요소를 삭제
// stack이 비어있는 상태인데 ')'일 경우 false
// stack이 비어있지 않다면 false
// 시간복잡도: O(n)

File: 프로그래머스/2/12909. 올바른 괄호/올바른 괄호.py
def solution(s):
    stack = []
    for i in s:
        if len(stack) == 0 and i == ')':
            return False
        elif i == '(':
            stack.append(s)
        elif i == ')':
            stack.pop()
        else:
            print("에러")
        
    if len(stack) != 0:
        return False
    return True
    
# stack을 유지
# '(': append
# ')': pop

# 1. len(stack) == 0인데 ')'이 들어오는 경우
# 2. loop를 끝냈는 데 len(stack) != 0인 경우

# 시간복잡도 O(n)
# 공간복잡도 O(n)